.github/workflows/ci.yml

```diff
@@ -1,39 +1,37 @@
 name: CI
-on:
-  push:
-    branches: [master]
-  pull_request:
-    branches: [master]
+on: [push, pull_request]
 
 jobs:
   test:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        mongodb: ["4.4"]
-        ruby: ["2.5", "2.6", "2.7", "3.0"]
-        gemfile:
-          - mongoid-3.1
-          - mongoid-4.0
-          - mongoid-5.0
-          - mongoid-6.0
-          - mongoid-7.0
-          - mongoid-8.0
-
+        entry:
+          - { ruby: "2.6", mongodb: '4.4', gemfile: 'mongoid-4.0', bundler: '2.0.0.pre.3' }
+          - { ruby: "2.7", mongodb: '4.4', gemfile: 'mongoid-5.0', bundler: '2.0.0.pre.3' }
+          - { ruby: "2.7", mongodb: '4.4', gemfile: 'mongoid-6.0' }
+          - { ruby: "2.7", mongodb: '4.4', gemfile: 'mongoid-7.0' }
+          - { ruby: "3.1", mongodb: '4.4', gemfile: 'mongoid-8.0' }
+    name: test (ruby=${{ matrix.entry.ruby }}, mongodb=${{ matrix.entry.mongodb }}), gemfile=${{ matrix.entry.gemfile }})
     env:
-      BUNDLE_GEMFILE: gemfiles/${{ matrix.gemfile }}.gemfile
+      BUNDLE_GEMFILE: gemfiles/${{ matrix.entry.gemfile }}.gemfile
     steps:
-      - name: Set up MongoDB ${{ matrix.mongodb }}
+      - name: Set up MongoDB ${{ matrix.entry.mongodb }}
         uses: supercharge/mongodb-github-action@1.8.0
         with:
-          mongodb-version: ${{ matrix.mongodb }}
-      # https://github.com/marketplace/actions/checkout
+          mongodb-version: ${{ matrix.entry.mongodb }}
       - uses: actions/checkout@v2
-      # https://github.com/marketplace/actions/setup-ruby-jruby-and-truffleruby
       - name: Set up Ruby
         uses: ruby/setup-ruby@v1
         with:
-          ruby-version: ${{ matrix.ruby }}
-          bundler-cache: true
+          ruby-version: ${{ matrix.entry.ruby }}
+      - name: Bundle install w/custom bundler
+        if: ${{ matrix.entry.bundler }}
+        run: |
+          gem install bundler:${{ matrix.entry.bundler }}
+          bundle _${{ matrix.entry.bundler }}_ install
+      - name: Bundle install with default bundler
+        if: ${{ !matrix.entry.bundler }}
+        run: bundle install
```

Comment:
[CONFUSED] `ruby/setup-ruby` supports a `bundler` option.

For example:

```yaml
      - name: Set up Ruby
        uses: ruby/setup-ruby@v1
        with:
          ruby-version: ${{ matrix.ruby }}
          bundler: ${{ matrix.bundler }}
```

Why the extra work here to specify the bundler version?  Also, for older ruby versions, just specify bundler 1.

Reply:
I cannot get the older rubies + rails to work with anything other than bundler 2.0.0.pre.3. So I chose the version recommended. That said, I think I can do `bundler: 2.0.0.pre.3`, I didn't realize that was a setting.